lib/auth/auth.go

```diff
@@ -6078,11 +6076,23 @@ func (a *Server) isMFARequired(ctx context.Context, checker services.AccessCheck
 			services.NewWindowsLoginMatcher(t.WindowsDesktop.GetLogin()))
 
 	case *proto.IsMFARequiredRequest_App:
-		app, err := a.GetApp(ctx, t.App.GetName())
+		if t.App.Name == "" {
+			return nil, trace.BadParameter("missing Name field in an app-only UserCertsRequest")
+		}
+
+		servers, err := a.GetApplicationServers(ctx, apidefaults.Namespace)
```

Comment:
Is it correct to say that this went from looking up a single app by name to fetching every app in the cluster?

Is there a more efficient way to look up the app rather than pulling back potentially many thousands of them?

Reply:
Could we at least use a paginated API and check for the presence of the app one page at a time?

If there are 20,000 apps and the one we're looking for is first in the list there's no reason why we should have to fetch and unmarshal the rest of them, right?

Reply:
@zmb3 Is this ok to merge if Tim plans on making a follow up PR?

Reply:
You bet!